### ISSUE TITLE
Stream Publishing prevents Grain Collection

### ISSUE THREAD

**Issue by brinkrob** [dotnet/orleans#1533]
I have a grain that publishes to a stream upon the callback of a 30 second timer.

My expectation is that the grain would get collected after 2 hours of no other activity since the timer itself is not keeping the grain alive. But the grain appears to live forever.

I've been able to isolate the issue to the publishing of the stream.

I changed the configuration to collect after 2 minutes.  When the timer callback publishes to the stream, the deactivation never occurs.  I also have breakpoints on all public methods to ensure that they are not being called.

When I removed the publishing of the stream from the timer callback, I was able to see the grain deactivate in ~2 minutes. (even though I saw the timer callback still was being executed)

This was found on 1.1.2.

**Comment by jason-bragg** [comment]
What stream provider?  SimpleMessagingStreamProvider?
Are you publishing to the same stream every 30 seconds or is it a different stream each time?
Is the grain subscribing to any streams, or only producing?

**Comment by brinkrob** [comment]
SimpleMessagingStreamProvider

It is publishing to the same stream every time.

It is not subscribed to any streams.

**Comment by jason-bragg** [comment]
TL;DR - Publishing an event on a stream is being recognized as grain activity.

I'll need to confirm with those that are more familiar with grain GC, but here's what it looks like to me..
Even though the call publishing the event is being done from a timer, which is outside the threading context of the grain, it is still a call from the grain.  Since Orleans grain calls follow an RPC style call pattern, all calls have responses, even if only to report errors or timeouts.  This response is being executed in the context of the grain, which is registered as grain activity, delaying deactivation.

I'm unclear if this is a bug, as the grain is not exactly idle.. it is doing stuff..

**Comment by sergeybykov** [comment]
>I'm unclear if this is a bug, as the grain is not exactly idle.. it is doing stuff..
 Though I concede, this behavior surprised me as well.

To be consistent, I think we need to treat all timer-originated activities the same way. If we don't treat simple timer calls as activities resetting activation's idle time, we shouldn't have a different semantic for when it publishes to a stream.

**Comment by jason-bragg** [comment]
> ..we shouldn't have a different semantic for when it publishes to a stream.

This is not specific to streams.  Any grain call made from a grain resets its idle time.

**Comment by gabikliot** [comment]
Agree on what Sergey wrote, we need consistent behaviour. Sounds like we should simply not treat replies as activities. Probably a one line fix, add if in the catalog before resetting the idle period for replies.

**Comment by jason-bragg** [comment]
From previous comment:
>I'm unclear if this is a bug, as the grain is not exactly idle.. it is doing stuff..

I'm not (yet) arguing a position as to what should be done (or not done), only considering the behaviors.  I agree that the behavior should be consistent, but that really depends on what we consider 'activity'.

Semantically, what is activity?  Timers don't wake grains, or run under their context, but if the grain makes other grain calls during a timer call, is that considered activity?  A grain must have an activation to make a grain call.  Does anything done in the context of a timer not count as activity?  What about loading/persisting state?  Registering reminders?  Subscribing to streams?

Again, I'm not making an argument here as to how we should handle this, I am just not yet ready to modify the existing behavior without a clear understanding of the expected semantics and thought through their implications.

As I understand it, currently an activation's activity is corolated with tasks running under it's context.  Any action that triggers execution under the grains context (including resolving responses from grain calls) is considered activity.

**Comment by sergeybykov** [comment]
In my opinion we should only consider activities originated from outside - calls *to* the grain or events *received* by the grain, but not outgoing calls of events.

**Comment by gabikliot** [comment]
Jason, what counts as activity for activation GC is described in the doc on io pages. Just look there, it is very well explained.

**Comment by jason-bragg** [comment]
@gabikliot, Thanks, that references the exact case that categorizes this as a bug.
Linking and quoting for convenience.  From [Activation Garbage Collection]( http://dotnet.github.io/orleans/Advanced-Concepts/Activation-Garbage-Collection)

>What counts as “being active” for the purpose of grain activation collection
• receiving a call
• receiving a reminder

>What does NOT count as “being active” for the purpose of grain activation collection
• performing a call (to another grain or to an Orleans client)
• timer events
• arbitrary IO operations or external calls not involving Orleans framework
